preformance enhanced quicksort

king-11 opened the pull request:
- Remove Tail Recursive Call 
- Save memory space by reducing call stack worst-case size

draft for #17178

king-11 commented:
current time metric averaging using following script
```chapel
module test_sort {
	use Sort.QuickSort;
	use Random;
	use Time;
	config const n = 1000000000;
	config const y = 10;
	var arr: [0..n] uint;


	var timer:Timer;
	timer.start();
	for x in 1..y {
		fillRandom(arr);
		quickSort(arr);
	}
	timer.stop();
	writeln("Sorted ", n, " elements in ", timer.elapsed()/y, " seconds");
	writeln(8*n/timer.elapsed()/1024.0/1024.0, " MiB/s");
}
```

king-11 commented:
@ronawho @gbtitus requesting your review.

I am done with the initial implementation of the changes that am suggesting.
Observed :
- the timer results aren't stable as am using a random array for each iteration but on average both the implementations are performing equally in terms of size and heap allocation (checked using [HeapTrack](https://github.com/KDE/heaptrack) )

Help:
It was trying to check the stack size using Valgrind but the program doesn't seem to complete any help on this part will allow checking whether the implementation is practically an improvement or not. 
  ```bash
  valgrind --tool=massif --stacks=yes --heap=no --time-unit=ms ./sort.out 
  ```

gbtitus commented:
Ah, actually now that I see the details of the code, tail recursion elimination shouldn't be expected to save any memory here. We are only eliminating the recursive step that sorts the high half (when there are fewer than 300 things to sort). The non-tail recursive step that sorts the low half still remains. But the stack frame for the tail recursion we eliminated would have resided in the same memory as the stack frame for the non-tail recursion that we didn't eliminate anyway, so there is no net effect on stack usage.

I'm not at all knowledgeable about valgrind, so I can't comment on its failure to complete other than to ask: are you sure it doesn't complete? Is it possible it's just running much more slowly that you expected? I realize I am asking you to solve the "halting problem".  :-)

One note about that valgrind run and the `--stacks=yes --heap=no` options I see there -- the memory blocks for Chapel task stacks are actually allocated out of the heap. Unless valgrind has special provision for treating our task stack blocks as stacks (@ronawho, does it?), they will look like heap allocations to it.

king-11 commented:
@gbtitus ahh got the point and I also saw my mistake in implementing this, the usual tail recursion elimination doesn't select the lower or upper half it selects the larger half for looping and smaller for recursion. 🤔 So if I implement this properly now will it provide the initial optimization that we were looking for ?

king-11 commented:
@gbtitus @bradcray require your further review on this.

I made the required changes to ensure that the stack usage is less now as we are calling the recursive step for the smaller partition which in the worst case will require memory of ```O(log(n))``` when then partitions are divided into equals. while the earlier implementation could have gone to ```O(n)``` recursions when the worst case of the quick sort would have occured.

king-11 commented:
@bradcray @gbtitus I was finally able to triage this update further with a new scripts to maintain consistency in both the functions instead of random array which I was using earlier and update the optimizations as well
```chapel
module write_into {
	use Random;
  use IO;
	config const n = 300000000;
  config const filename = "./arr.txt";
	config const y = 1;
	var arr: [0..n] int;
  var outfile = try! open(filename,iomode.cw);
  var writer = try! outfile.writer();

	fillRandom(arr);
  try! writer.write(arr);
  try! writer.close();
  try! outfile.close();
}
```
```chapel
module test_sort {
  use IO;
  use Sort.QuickSort;
  use Time;
  config const n = 300000000;
  config const filename = "./arr.txt";
	config const y = 1;
	var arr: [0..n] int;
  var outfile = try! open(filename,iomode.r);
  var reader = try! outfile.reader();

  try! reader.read(arr);
  try! reader.close();
  try! outfile.close();
  var timer:Timer;
  timer.start();
  quickSort(arr);
	timer.stop();
	writeln("Sorted ", n, " elements in ", timer.elapsed(), " seconds");
	writeln(32*n/timer.elapsed()/1024.0/1024.0, " MiB/s");
}
```
Observations :
- I consistently under the same runtime and compile-time conditions found that I newer implementation performs better than the current one in terms of time. i.e. on avg `25` secs vs `27` secs.
- The memory part that earlier I wasn't able to clarify on much can be explained now. The tail elimination is done for the larger recursive call which highly reduces the number of total recursive calls made even though the smaller one exists in the same frame as @gbtitus mentioned the recursive calls it will make will be less and hence the memory reduction part.
- Earlier I was only removing the lower part of the call which wasn't any optimization but with the conditional check, the optimization is introduced now.

bradcray commented:
Hi @king-11 — Sorry for the lack of a response here.  We've had a busy few weeks.  I'm tagging @mppf on this, as he's done most of the most recent design and implementation work in the Sort module to get his thoughts on this effort.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -838,14 +838,14 @@ module TimSort {
     /* apply merge operations on each block
     *as the merges at a level are independent of each other
-    *they can be applied parallelly
+    *they can be applied parallely
```
it should say "they can be applied in parallel"

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -1233,52 +1233,62 @@ module QuickSort {
-      return;
-    }
+    var lo = start,
+          hi = end;
```
I think we need to remove two leading spaces on this line so that the 'lo' and 'hi' are in the same columns

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -1233,52 +1233,62 @@ module QuickSort {
+
+        if hi-lo < 300 {
+          // stay sequential
+          if(eqStart - lo > hi - eqEnd){
```
We don't need to put `( )` around the condition of an `if` statement in Chapel

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -1233,52 +1233,62 @@ module QuickSort {
+          if(eqStart - lo > hi - eqEnd){
+            quickSortImpl(Data, minlen, comparator, eqEnd+1, hi);
+            hi = eqStart - 1;
+          } else {
```
Now that this code has an `if`/`else` it is only doing half of the required quick sorting. It is my understanding that implementations of quick sort that do not use a recursive call need to have a manual "stack" of regions of the array that need to be sorted.

I don't think that what you have here is not doing tail call optimization. Rather it is changing the algorithm. Is there some prior work describing why this algorithm change would be a good idea? It does not look right to me.

Certainly it is possible to directly do one of the two subproblems and put the other one on a stack for later work, but that's not what this does.

I would expect that if we change it to use a stack of subproblems, it could do things like sort that stack of subproblems by size in order to have better parallel load balance.

king-11 replied:
>Now that this code has an if/else it is only doing half of the required quick sorting.

I ran a few tests, it actually completes both subproblems but the larger half is solved in the current task/function call only whereas for smaller we make a recursive call the role of `while (lo <= hi)` and `hi = eqStart - 1` or `lo = eqEnd + 1` ensures that this work is performed for the left out the problem as well. 

I wasn't able to find any resource for tail call elimination for three-way quick sort so I carried over the idea from two way quicksort from the following [resource](https://www.geeksforgeeks.org/quicksort-tail-call-optimization-reducing-worst-case-space-log-n/)

>It is my understanding that implementations of quicksort that do not use a recursive call need to have a manual "stack" of regions of the array that need to be sorted.

The stack ones that you mentioned we can use but will be a completely iterative solution which will be more complex to implement as per a paper I read about load balancing in parallel but performance gain will be good but currently, I was just trying to solve just the tail call optimization problem.

mppf replied:
Ah, I see in your linked resource, a comment like this:

>         // If left part is smaller, then recur for left
>         // part and handle right part iteratively

This is the part that I was missing. I think what you have is probably fine but we need to add a comment about how both the left and right parts are handled; one iteratively and one in a recursive call. I don't see any problem extending this idea to the 3-way quicksort (we just leave the equal elements alone).

king-11 replied:
Okay I will add few more documentation comments to make the idea more clear

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -1233,52 +1233,62 @@ module QuickSort {
-      // median of the medians
-      piv = order3(Data, medLo, medMid, medHi, comparator);
-    }
+    while(lo <= hi){
```
If there's just one element, it's in order, right? So this can be `while lo < hi {`

king-11 replied:
Yeah missed this

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -1233,52 +1233,67 @@ module QuickSort {
+        var (eqStart, eqEnd) = partition(Data, lo, piv, hi, comparator);
+
+        // stay sequential
+        if hi-lo < 300 {
```
Now the improvement is limited to code that runs only with 300 or fewer elements. But, the `forall` statement below is just checking to see if there are already enough tasks running to keep the cores busy. It might be reasonable to change this branch to say something like `if hi-lo < 300 || runningEnoughParallelTasks {`. The tricky bit is to figure out what exactly to put for `runningEnoughParallelTasks` / how to compute that (where I know of a not-user-facing way to do that but I'm not sure if it's reasonable to do here). Anyway this could be done in another PR.

king-11 replied:
oh, that will be interesting to deal with, I will try to look out some method if am able to I will open a PR for the same :)

mppf replied:
Aha, you can call `here.runningTasks()`. See https://chapel-lang.org/docs/master/usingchapel/tasks.html#task-related-quantification-methods-on-locales . You can use something like this code:

https://github.com/chapel-lang/chapel/blob/d690e1b53738157f4f5000df9d7c8d08ad4d05c9/modules/internal/ChapelRange.chpl#L2232-L2240

Do you want to try to do it in this PR so we can try for more performance improvement when we measure it?

king-11 replied:
yeah I will try it out pushing the changes now

king-11 replied:
Not sure although if I implemented it the right way or not 😅

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -1233,52 +1233,71 @@ module QuickSort {
-          quickSortImpl(Data, minlen, comparator, eqEnd+1, hi);
+        var (eqStart, eqEnd) = partition(Data, lo, piv, hi, comparator);
+        const tasksPerLocale = dataParTasksPerLocale;
+        const ignoreRunning = dataParIgnoreRunningTasks;
```
I think for quicksort's purposes, we should assume `ignoreRunning=false`.

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -1233,52 +1233,71 @@ module QuickSort {
-        else
-          quickSortImpl(Data, minlen, comparator, eqEnd+1, hi);
+        var (eqStart, eqEnd) = partition(Data, lo, piv, hi, comparator);
+        const tasksPerLocale = dataParTasksPerLocale;
```
Let's move this variable as well as `dptpl` outside of the while loop

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -1233,52 +1233,71 @@ module QuickSort {
+        const ignoreRunning = dataParIgnoreRunningTasks;
+        const otherTasks = if ignoreRunning then 0 else here.runningTasks() - 1;
+        const dptpl = if tasksPerLocale == 0 then here.maxTaskPar else tasksPerLocale;
+        const runningEnoughParallelTasks = if otherTasks < dptpl then false else true;
```
`        const runningEnoughParallelTasks = otherTasks >= dptpl;`

mppf commented on modules/packages/Sort.chpl:
```diff
@@ -1233,52 +1233,71 @@ module QuickSort {
+          }
+        } else {
+          // do the subproblems in parallel
+          forall i in 1..2 {
```
Now that we have `runningEnoughParallelTasks`, this section should just be

``` chapel
cobegin {
  quickSortImpl(Data, minlen, comparator, lo, eqStart-1);
  quickSortImpl(Data, minlen, comparator, eqEnd+1, hi);
}
```

king-11 commented:
@mppf I am not familiar with the error that make_mason is encountering I tried running manual checks the method seems to be working can you please take a look thanks

mppf commented:
hi @king-11 yes that is a strange error. I'm able to reproduced by checking out your branch and then running `make mason`. I think it's a compiler bug though. Please cherry-pick 760e6e23f47b56941412a73fafa3fd23176c5a16 for a workaround to the mason code. Meanwhile, I've created this issue #17623 to track the compiler bug.

mppf commented:
> I consistently under the same runtime and compile-time conditions found that I newer implementation performs better than the current one in terms of time. i.e. on avg 25 secs vs 27 secs.

Can you check the performance impact once more now that we are getting closer to the version to merge? Thanks.

king-11 commented:
@mppf I added one more commit which made the performance better it was based on the implementation of merge sort
https://github.com/chapel-lang/chapel/blob/df492363e3f4f9adfc47b774a4827241a4caa75c/modules/packages/Sort.chpl#L991
![all](https://user-images.githubusercontent.com/52061363/117165765-69414a00-ade3-11eb-9752-bf79826dc988.png)


- First one is using master
- Second is without the last commit
- the last one is after adding the last commit

One more thing I tried out was adding #17279 and that also added to the performance

mppf commented:
@king-11 sounds good. Do you think this PR is ready for testing?

Only other feedback I have is to ask - Does it make sense to move the `here.numPUs` out of the while loop? I don't know offhand how expensive this call is. If it's very expensive, we'd want to also remove it from the recursive calls, and pass the result as an argument. But if it does nothing for performance it's probably better to leave it the way you have it now since it is simpler.

king-11 commented:
@mppf I tried running the performance test by switching the `here.numPUs()` outside the loop it didn't make any performance change that could be consistently observed so I guess the function call isn't that costly. and yeah the pr is ready for testing

king-11 commented:
the tests that were failing in #17279 are still failing 🤔

mppf commented:
hi @king-11 - let's leave the minlen condition change to PR #17279. Can you take that commit out of this one?

king-11 commented:
@mppf made the change

mppf commented:
OK, great. I'm going to do a testing run.

mppf commented:
OK, this PR has passed testing. 

@ben-albrecht - this PR includes some Mason changes that I developed to work around issue #17623. Would you be able to review the Mason part at least? Thanks.

ben-albrecht reviewed (approved):
Mason changes look good. It'd be good to confirm `test/mason` passes if we already didn't. Thanks!
